go-controller/pkg/ovn/dns-name-resolver/egressfirewall_external_dns.go

```diff
@@ -0,0 +1,356 @@
+package dnsnameresolver
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"reflect"
+	"sync"
+	"time"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	ocpnetworkapiv1alpha1 "github.com/openshift/api/network/v1alpha1"
+	ocpnetworklisterv1alpha1 "github.com/openshift/client-go/network/listers/network/v1alpha1"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/controller"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	addressset "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/address_set"
+)
+
+// ExternalEgressDNS keeps track of DNS names and the corresponding IP addresses.
+// For each DNS name, an address set is allocated and the address set is
+// kept updated with the corresponding IP addresses. Whenever a DNS name
+// is removed the corresponding addresset is destroyed.
+type ExternalEgressDNS struct {
+	// dnsLock protects the operations performed on the maps related to
+	// the address sets.
+	dnsLock sync.Mutex
+	// dnsNames maps the dnsNames to the dnsResolvedNames.
+	dnsNames map[string]*dnsResolvedName
+	// namespaceDNSNames maps the DNS names used in the EgressFirewall rule of
+	// a namespace to the namespace
+	namespaceDNSNames map[string]sets.Set[string]
+	// addressSetFactory helps in the creation of the address sets.
+	addressSetFactory addressset.AddressSetFactory
+	// controllerName gives the name of the controller which created this
+	// instance of the ExternalEgressDNS.
+	controllerName string
```

Comment:
I was a bit lost with two sets of internal structs protected by separate locks, so maybe we can make this part its own struct (together with the methods changing it)? Call it something like `dnsTracker`, then just have an instance of `dnsTracker` in `ExternalEgressDNS`.
Then `ExternalEgressDNS` will only propagate Add and Delete, and have its own `reconcileDNSNameResolver`

Reply:
I have moved the fields of the struct related to the controller to a separate `dnsnameresolver-controller.go` file and created a `dnsNameResolverController` struct containing those fields.